# Amazon Products Data Analysis

## Data Cleaning & Preprocessing

### Analyst: Raneem jonaid

### Date: April 29, 2026


## Step 1: Loading the Data
In this step we load the raw Amazon products dataset from a pickle file.

In [1]:
# Import required libraries
import pickle  # To load the pkl file
import pandas as pd  # To work with the data

# Load the dataset
with open('Amazon_products.pkl', 'rb') as file:
    data = pickle.load(file)

# Confirm data loaded correctly
print("Data loaded successfully!")
print("Shape:", data.shape)


Data loaded successfully!
Shape: (1242, 16)


## Step 2: Data Quality Check
Checking missing values across all columns.

In [2]:
# Check missing values in all columns
print("Missing values per column:")
print(data.isnull().sum())

Missing values per column:
asin                              0
title                             0
price                            70
list_price                      585
rating                           18
reviews                          18
sold_past_month                 108
is_bestseller                     0
is_prime                          0
is_amazon_choice                  0
has_sustainability_features       0
available_offers                467
amazon_choice_type             1173
brand                           903
free_delivery_date              654
fastest_delivery_date           920
dtype: int64


## Step 3: Brand Cleaning
903 products have missing brand names. We will extract brands from product titles using one method:
- Method 1: Extract first word if it matches a known brand
- Products with no identifiable brand will be labeled as 'Generic'tle

### Method 1: Extract First Word from Title
First we extract the first word from titles with missing brands to identify potential brand names.

In [3]:
# Create a copy to work with
data_filled = data.copy()

# Extract first word from titles with missing brands
missing_brand_data = data_filled[data_filled['brand'].isnull()].copy()
missing_brand_data['extracted_brand'] = missing_brand_data['title'].str.split().str[0]

# Show all unique first words
unique_first_words = sorted(missing_brand_data['extracted_brand'].unique())
print(f"Total unique first words: {len(unique_first_words)}")
print("\nAll unique first words:")
print(unique_first_words)

Total unique first words: 239

All unique first words:
['(2', '15', '15.6', "15.6''", '15.6inch', '16.0-inch', '1InTheOffice', '2', '2020', '2021', '2023', '24', '24.5', '25', '27', '2pcs', '32', '34', '3M', '45W', '55°', '65%', '65W', '8', '90W', 'ACEMAGIC', 'ANTEMPER', 'AOC', 'ARZOPA', 'ASUS', 'AULA', 'AYY', 'AboveTEK', 'Acer', 'Acrylic', 'Adjustable', 'Aelfox', 'Aesthetic', 'Allsop', 'Amazon', 'Anker', 'Apple', 'Arteck', 'BAGSMART', 'BANGE', 'BENGOO', 'BESIGN', 'BIGASUO', 'BRILA', 'Backlit', 'Beautiful', 'Belkin', 'BenQ', 'Blue', 'Bluetooth', 'CRUA', 'Cat', 'Charger', 'Cloud', 'Comfy', 'Compressed', 'Computer', 'Contour', 'Corsair', 'Cute', 'DELL', 'DELTAHUB', 'DIERYA', 'Dapesuom', 'Dell', 'Dumpster', 'E-YOOSO', 'ECHTPower', 'EMPSIGN', 'EMSHOI', 'EPOMAKER', 'EVGA', 'Emetok', 'Ergonomic', 'Excel', 'Fellowes', 'Five', 'Foldable', 'Forvencer', 'G', 'GAMEPOWER', 'GIM', 'Gaming', 'Geekria', 'Gel', 'Gimars', 'Glider', 'Glorious', 'Gorilla', 'Green', 'HP', 'HUANUO', 'Hokafenle', 'HornsX1',

In [4]:
known_brands = [
    # Laptops & Computer Brands
    'ACEMAGIC', 'Acer', 'AOC', 'ARZOPA', 'ASUS', 'Apple', 'Dell', 'DELL',
    'ECHTPower', 'EMPSIGN', 'EMSHOI', 'HP', 'HyperX', 'INNOCN', 'Lenovo',
    'LG', 'MNN', 'MSI', 'Microsoft', 'NewAcer', 'SAMSUNG', 'SANSUI', 'Sceptre',
    'SenseAGE', 'ViewSonic', 'VisionTek', 'Yodoit',
    # Monitors
    'BenQ', 'KOORUI', 'KTC', 'PHILIPS', 'Vaydeer', 'VisionOwl',
    # Keyboards & Mouse
    'AULA', 'AYY', 'BENGOO', 'Corsair', 'DIERYA', 'E-YOOSO', 'EPOMAKER',
    'EVGA', 'GAMEPOWER', 'GIM', 'Gimars', 'Glorious', 'HornsX1', 'ITNRSIIET',
    'JIKIOU', 'KLIM', 'Logitech', 'logitech', 'MageGee', 'MEETION', 'MROCO',
    'Nulea', 'OMOTON', 'Perixx', 'ProtoArc', 'QIYI', 'RAINSMORE',
    'Razer', 'Redragon', 'Rii', 'SOUNDANCE', 'SteelSeries', 'TECKNET',
    'VANKEY', 'Weewooday', 'Womier', 'YEYIAN', 'YUNZII', 'YXLILI',
    'Ytonet', 'Yunsailing', 'ZZA', 'havit', 'iClever', 'seenda', 'tounee',
    'VILVA', 'WGK', 'YSAGi', 'VssoPlor', 'Snpurdiri', 'Omhuzsyo', 'Taygeer',
    'MEKASS', 'Emetok', 'Hokafenle', 'SABLUTE', 'SHRRADOO', 'Dapesuom',
    # Accessories & Stands
    'AboveTEK', 'Allsop', 'ANTEMPER', 'Arteck', 'BAGSMART', 'BANGE',
    'BESIGN', 'BIGASUO', 'BRILA', 'Belkin', 'Contour', 'DELTAHUB',
    'Fellowes', 'Geekria', 'Gorilla', 'HUANUO', 'Kensington', 'KTRIO',
    'LAPGEAR', 'Lamicall', 'Lekvey', 'LOVEVOOK', 'Lovaky', 'MATEIN',
    'Manhattan', 'MOSISO', 'OXO', 'OTVOC', 'PAPERAGE', 'Targus', 'TOROKOM',
    'UPPERCASE', 'VANKEAN', 'Verbatim', 'Anker',
    # Notebooks & Stationery
    'Five Star', 'Forvencer', 'KOKUYO', 'LEUCHTTURM1917', 'LEUCHTTURM1917-120G',
    'Maruman', 'Mead', 'Midori', 'Moleskine', 'MUJI', 'Oxford',
    'Rocketbook', 'SUIN', 'Yansanido', 'Z-Edge', 'ZERU', 'ZICOTO', 'Zibly',
    # Other
    'Amazon Basics', 'CRUA', 'KYY', 'SUNEE', 'iCasso', 'Aelfox', 'RisoPhy',
]

print(f"Total known brands: {len(known_brands)}")

Total known brands: 152


### Step 3.2: Fill Missing Brands Using First Word
Fill missing brands where the first word of the title matches a known brand.

In [5]:
# Roll back and use ONLY Method 1
data_filled = data.copy()

# Method 1 ONLY: First word extraction
for idx, row in data_filled[data_filled['brand'].isnull()].iterrows():
    first_word = row['title'].split()[0]
    if first_word in known_brands:
        data_filled.loc[idx, 'brand'] = first_word

print(f"Missing brands before: {data['brand'].isnull().sum()}")
print(f"Missing brands after Method 1: {data_filled['brand'].isnull().sum()}")
print(f"Brands filled: {data['brand'].isnull().sum() - data_filled['brand'].isnull().sum()}")

Missing brands before: 903
Missing brands after Method 1: 236
Brands filled: 667


### Step 3.3: Fill Remaining Missing Brands as Generic
Products with no identifiable brand name in the title are labeled as 'Generic'.

In [6]:
# Fill remaining missing brands with 'Generic'
data_filled.loc[data_filled['brand'].isnull(), 'brand'] = 'Generic'

print("Brand Cleaning Summary:")
print(f"Total products: {len(data_filled)}")
print(f"Brands filled from Method 1: 667")
print(f"Filled as Generic: 236")
print(f"Missing brands remaining: {data_filled['brand'].isnull().sum()}")

Brand Cleaning Summary:
Total products: 1242
Brands filled from Method 1: 667
Filled as Generic: 236
Missing brands remaining: 0


In [7]:
data.head(60)

,asin,title,price,list_price,rating,reviews,sold_past_month,is_bestseller,is_prime,is_amazon_choice,has_sustainability_features,available_offers,amazon_choice_type,brand,free_delivery_date,fastest_delivery_date
0,B0DJK7NW1J,"15.6 Inch Laptops, Windows 11 Laptop Computers...",199.98,679.99,4.4 out of 5 stars,48,100+,False,False,False,False,NaN,NaN,NaN,NaN,NaN
1,B0BS4BP8FB,Acer Aspire 3 A315-24P-R7VH Slim Laptop | 15.6...,279.99,321.99,4.4 out of 5 stars,"39,243",8K+,False,False,True,False,6,Overall Pick,NaN,NaN,NaN
2,B0DKDY78K3,"Newest Gaming Laptop, Laptop with AMD Ryzen 7 ...",649.99,"1,699.99",4.8 out of 5 stars,18,100+,False,False,False,False,NaN,NaN,NaN,NaN,NaN
3,B0CPL25J3W,"HP Portable Laptop, Student and Business, 14"" ...",197.35,269.00,4.1 out of 5 stars,"1,678",1K+,False,False,False,False,25,NaN,NaN,NaN,NaN
4,B0947BJ67M,"HP 14 Laptop, Intel Celeron N4020, 4 GB RAM, 6...",176.00,209.99,4.0 out of 5 stars,"1,861",4K+,False,False,False,False,50,NaN,NaN,NaN,NaN
5,B0B2D77YB8,"HP Newest 14"" Ultral Light Laptop for Students...",249.99,NaN,4.0 out of 5 stars,"1,950",2K+,False,False,False,False,NaN,NaN,NaN,NaN,NaN
6,B0DK9HYLFP,"ACEMAGIC Laptop, Newest Windows 11 Laptop Comp...",359.99,"1,299.99",4.8 out of 5 stars,24,100+,False,False,False,False,NaN,NaN,NaN,NaN,NaN
7,B0DHXKL1BF,"HP Flagship 14"" Laptop with 1 Year Office 365 ...",339.00,NaN,4.3 out of 5 stars,13,100+,False,False,False,False,NaN,NaN,NaN,NaN,NaN
8,B09YRY6QCX,ASUS E410 Intel Celeron N4020 4GB 64GB 14-Inch...,147.00,NaN,4.1 out of 5 stars,678,500+,False,False,False,False,NaN,NaN,NaN,NaN,NaN
9,B0DFLHNVKM,"16.0-inch Laptop, 16GB DDR4 512GB SSD, Newest ...",NaN,NaN,3.5 out of 5 stars,6,500+,False,False,False,False,NaN,NaN,NaN,NaN,NaN



## Step 4: Handling Duplicates
Before handling missing values, we first check if any products 
appear more than once in the dataset. If a duplicate has a value 
that another duplicate is missing, we can use it to fill the gap.

### Step 4.1: Check for Duplicates
We check how many duplicate rows exist in the dataset.
A duplicate row is an exact copy of another row.

In [8]:
# Check for duplicate products
print("Total rows:", len(data_filled))
print("Duplicate rows:", data_filled.duplicated().sum())
print()



Total rows: 1242
Duplicate rows: 127



### Step 4.2: Count How Many Times Each Product Appears
We group duplicate products by title and count their occurrences
to understand how many times each product is repeated.

In [9]:
# Count occurrences of each title
title_counts = data_filled[data_filled.duplicated(subset=['title'], keep=False)].groupby('title').size().reset_index()
title_counts.columns = ['title', 'count']
title_counts = title_counts.sort_values('count', ascending=False)
print(title_counts)



                                                 title  count
143  TECKNET Wireless Mouse, 2.4G Ergonomic Optical...      6
130  Sceptre 20" 1600x900 75Hz Ultra Thin LED Monit...      5
94   Logitech M185 Wireless Mouse, 2.4GHz with USB ...      4
96   Logitech M510 Wireless Computer Mouse for PC w...      4
99   Logitech MX Keys S Wireless Keyboard, Low Prof...      4
..                                                 ...    ...
54   Chromebook CM14 Laptop, 14" HD Anti-Glare Disp...      2
53   CRUA 27" 144Hz/180Hz Curved Gaming Monitor, FH...      2
52   CRUA 24" Curved Monitor, FHD(1920×1080P) 100HZ...      2
51   Bluetooth and 2.4G (2-in-1) Rechargeable Wirel...      2
153  acer Nitro 34" UWQHD 3440 x 1440 1000R Curved ...      2

[154 rows x 2 columns]


### Step 4.3: Fill Missing Values Using Duplicate Products
If a product appears multiple times and one entry has a missing value 
while another entry has the actual value, we fill the missing value 
from its duplicate.

#### Columns we attempt to fill:
- price
- rating
- reviews
- sold_past_month

In [10]:
# Fill missing values using duplicate products
cols_to_fill = ['price', 'rating', 'reviews', 'sold_past_month']

for col in cols_to_fill:
    # Group by title and fill NaN with first available value
    data_filled[col] = data_filled.groupby('title')[col].transform(
        lambda x: x.fillna(x.dropna().iloc[0] if len(x.dropna()) > 0 else x.iloc[0])
    )

# Check results
print("Missing values after filling from duplicates:")
print(data_filled[cols_to_fill].isnull().sum())

C:\Users\berri\AppData\Local\Temp\ipykernel_31740\3185954704.py:7: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  lambda x: x.fillna(x.dropna().iloc[0] if len(x.dropna()) > 0 else x.iloc[0])
C:\Users\berri\AppData\Local\Temp\ipykernel_31740\3185954704.py:7: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  lambda x: x.fillna(x.dropna().iloc[0] if len(x.dropna()) > 0 else x.iloc[0])
C:\Users\berri\AppData\Local\Temp\ipykernel_31740\3185954704.py:7: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a futu

Missing values after filling from duplicates:
price               65
rating              18
reviews             18
sold_past_month    108
dtype: int64


### Step 5.1: Drop Rows with Missing Price
Price is essential for our analysis. 
We drop the 65 rows where price is missing.

In [11]:
# dropna() removes rows where price is NaN
# subset=['price'] tells it to only check the price column
before = len(data_filled)
data_filled = data_filled.dropna(subset=['price'])
after = len(data_filled)

print(f"Rows before: {before}")
print(f"Rows after: {after}")
print(f"Rows dropped: {before - after}")

Rows before: 1242
Rows after: 1177
Rows dropped: 65


### Step 5.2: Remove Duplicate Rows
Keep only one copy of each duplicated product.

In [12]:
# drop_duplicates() removes rows with same title
# keeping only the first occurrence
before = len(data_filled)
data_filled = data_filled.drop_duplicates(subset=['title'])
after = len(data_filled)

print(f"Rows before: {before}")
print(f"Rows after: {after}")
print(f"Duplicates removed: {before - after}")

Rows before: 1177
Rows after: 971
Duplicates removed: 206


Step 5.3 - Verify final dataset 


In [13]:
# Final check to confirm our dataset is clean
print(f"Final dataset size: {len(data_filled)}")
print()
print("Missing values:")
print(data_filled.isnull().sum())
print()
print("Duplicates remaining:", data_filled.duplicated().sum())

Final dataset size: 971

Missing values:
asin                             0
title                            0
price                            0
list_price                     448
rating                          10
reviews                         10
sold_past_month                 86
is_bestseller                    0
is_prime                         0
is_amazon_choice                 0
has_sustainability_features      0
available_offers               387
amazon_choice_type             925
brand                            0
free_delivery_date             502
fastest_delivery_date          701
dtype: int64

Duplicates remaining: 0


## Step 6: Fix Data Types
Ensure all columns have correct data types before exporting to SQL.

In [14]:
# Check current data types
print("Current data types:")
print(data_filled.dtypes)
print()
print("Sample values:")
print(data_filled[['price', 'rating', 'reviews', 'sold_past_month']].head())

Current data types:
asin                           object
title                          object
price                          object
list_price                     object
rating                         object
reviews                        object
sold_past_month                object
is_bestseller                    bool
is_prime                         bool
is_amazon_choice                 bool
has_sustainability_features      bool
available_offers               object
amazon_choice_type             object
brand                          object
free_delivery_date             object
fastest_delivery_date          object
dtype: object

Sample values:
    price              rating reviews sold_past_month
0  199.98  4.4 out of 5 stars      48            100+
1  279.99  4.4 out of 5 stars  39,243             8K+
2  649.99  4.8 out of 5 stars      18            100+
3  197.35  4.1 out of 5 stars   1,678             1K+
4  176.00  4.0 out of 5 stars   1,861             4K+


### Step 6.1: Fix Data Types
Convert price, rating, reviews and sold_past_month from text to numbers.

In [15]:
# price: convert to float
data_filled['price'] = pd.to_numeric(data_filled['price'], errors='coerce')

# reviews: remove commas and convert to number
data_filled['reviews'] = data_filled['reviews'].str.replace(',', '').str.strip()
data_filled['reviews'] = pd.to_numeric(data_filled['reviews'], errors='coerce')

# Check results
print("Data types after fixing:")
print(data_filled[['price', 'rating', 'reviews']].dtypes)
print()
print("Sample values:")
print(data_filled[['price', 'rating', 'reviews']].head())

Data types after fixing:
price      float64
rating      object
reviews    float64
dtype: object

Sample values:
    price              rating  reviews
0  199.98  4.4 out of 5 stars     48.0
1  279.99  4.4 out of 5 stars  39243.0
2  649.99  4.8 out of 5 stars     18.0
3  197.35  4.1 out of 5 stars   1678.0
4  176.00  4.0 out of 5 stars   1861.0


In [16]:
# Check current data types and sample values
print("Current data types:")
print(data_filled[['price', 'rating', 'reviews', 'sold_past_month']].dtypes)
print()
print("Sample values:")
print(data_filled[['price', 'rating', 'reviews', 'sold_past_month']].head())

Current data types:
price              float64
rating              object
reviews            float64
sold_past_month     object
dtype: object

Sample values:
    price              rating  reviews sold_past_month
0  199.98  4.4 out of 5 stars     48.0            100+
1  279.99  4.4 out of 5 stars  39243.0             8K+
2  649.99  4.8 out of 5 stars     18.0            100+
3  197.35  4.1 out of 5 stars   1678.0             1K+
4  176.00  4.0 out of 5 stars   1861.0             4K+


### Step 6.2: Fix sold_past_month Column
Convert sold_past_month from text to numbers.
- "1K+" becomes 1000
- "8K+" becomes 8000
- "100+" becomes 100

In [17]:
# Convert sold_past_month to numbers
# 1K+ = 1000, 8K+ = 8000, 100+ = 100
def convert_sold(value):
    if pd.isna(value):
        return None
    value = str(value).replace('+', '').strip()
    if 'K' in value:
        return float(value.replace('K', '')) * 1000
    else:
        return float(value)

data_filled['sold_past_month'] = data_filled['sold_past_month'].apply(convert_sold)

print("Data types now:")
print(data_filled[['price', 'rating', 'reviews', 'sold_past_month']].dtypes)
print()
print("Sample values:")
print(data_filled[['price', 'rating', 'reviews', 'sold_past_month']].head())

Data types now:
price              float64
rating              object
reviews            float64
sold_past_month    float64
dtype: object

Sample values:
    price              rating  reviews  sold_past_month
0  199.98  4.4 out of 5 stars     48.0            100.0
1  279.99  4.4 out of 5 stars  39243.0           8000.0
2  649.99  4.8 out of 5 stars     18.0            100.0
3  197.35  4.1 out of 5 stars   1678.0           1000.0
4  176.00  4.0 out of 5 stars   1861.0           4000.0


## Step 7: Export Clean Data to CSV
We drop columns that are not needed for our analysis before exporting.

### Dropped columns:
- free_delivery_date: not needed for analysis
- fastest_delivery_date: not needed for analysis
- amazon_choice_type: not needed for analysis
- available_offers: not needed for analysis
- list_price: not needed for analysis

### Remaining columns will be used to answer our 3 business problems:
1. Pricing Strategy
2. Best Selling Products
3. Customer Satisfaction

In [29]:
# Drop columns not needed for our analysis
# data_filled = data_filled.drop(columns=[
#     'free_delivery_date', 
#     'fastest_delivery_date', 
#     'amazon_choice_type',
#     'available_offers',
#     'list_price'
# ])

print(f"Remaining columns: {data_filled.columns.tolist()}")

Remaining columns: ['asin', 'title', 'price', 'rating', 'reviews', 'sold_past_month', 'is_bestseller', 'is_prime', 'is_amazon_choice', 'has_sustainability_features', 'brand']


## Step 8: Export Clean Data to CSV
Export the cleaned dataset to a CSV file for importing into SQL Server.

In [32]:
# Export
data_filled.to_csv(r'C:\Users\berri\OneDrive\project portfolio\Amazon_product_Analysis\Data\clean_data\amazon_products_clean.csv', index=False)

print("Exported successfully!")

print(f"File: amazon_products_clean.csv")
print(f"Total rows: {len(data_filled)}")
print(f"Total columns: {len(data_filled.columns)}")




Exported successfully!
File: amazon_products_clean.csv
Total rows: 971
Total columns: 11
